lib/redis_mutex/lock.ex

```diff
@@ -1,118 +1,131 @@
     end
-    ```
-  """
-  defmacro with_lock(key, timeout \\ @default_timeout, expiry \\ @default_expiry, do: clause) do
-    quote do
-      key = unquote(key)
-      timeout = unquote(timeout)
-      expiry = unquote(expiry)
-      uuid = Uniq.UUID.uuid1()
+    """
+
+    @name RedisMutex
 
-      RedisMutex.Lock.take_lock(key, uuid, timeout, expiry)
+    @callback child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
 
-      block_value = unquote(clause)
+    @callback start_link(start_options :: RedisMutex.start_options()) ::
+                {:ok, pid()} | {:error, any()}
 
-      RedisMutex.Lock.unlock(key, uuid)
+    @callback with_lock(key :: String.t(), do: clause :: term()) :: any()
 
-      block_value
+    @callback with_lock(key :: String.t(), timeout :: integer(), do: clause :: term()) :: any()
+
+    @callback with_lock(key :: String.t(), timeout :: integer(), expiry :: integer(),
+                do: clause :: term()
+              ) ::
+                any()
+
+    @spec child_spec(Keyword.t()) :: Supervisor.child_spec()
+    def child_spec(opts) do
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, opts},
+        type: :worker,
+        restart: :permanent,
+        shutdown: 500
+      }
     end
-  end
 
-  @doc """
-  This function takes in a key, unique string, and a timeout in milliseconds.
-  It will call itself recursively until it is able to set a lock
-  or the timeout expires.
-  """
-  def take_lock(key, uuid, timeout \\ @default_timeout, expiry \\ @default_expiry, finish \\ nil)
+    @spec start_link(RedisMutex.start_options()) :: :ignore | {:error, any} | {:ok, pid}
+    def start_link(start_options \\ []) do
+      {redis_url, redix_opts} = Keyword.pop(start_options, :redis_url)
 
-  def take_lock(key, uuid, timeout, expiry, nil) do
-    finish = DateTime.add(DateTime.utc_now(), timeout, :millisecond)
-    take_lock(key, uuid, timeout, expiry, finish)
-  end
+      case redis_url do
+        redis_url when is_binary(redis_url) ->
+          Redix.start_link(redis_url, name: @name, sync_connect: true)
 
-  def take_lock(key, uuid, timeout, expiry, finish) do
-    if DateTime.compare(finish, DateTime.utc_now()) == :lt do
-      raise RedisMutex.Error, message: "Unable to obtain lock."
+        nil ->
+          [name: @name, sync_connect: true]
+          |> Keyword.merge(redix_opts)
+          |> Redix.start_link()
+      end
     end
 
-    if !lock(key, uuid, expiry) do
+    @spec with_lock(String.t(), integer(), integer(), do: term()) :: any()
+    defmacro with_lock(key, timeout \\ @default_timeout, expiry \\ @default_expiry, do: clause) do
+      quote do
+        key = unquote(key)
+        timeout = unquote(timeout)
+        expiry = unquote(expiry)
+        uuid = Uniq.UUID.uuid1()
+
+        RedisMutex.Lock.take_lock(key, uuid, timeout, expiry)
+
+        block_value = unquote(clause)
+
+        RedisMutex.Lock.unlock(key, uuid)
+
+        block_value
+      end
+    end
```

Comment:
Does this need to be a macro? Could it take a function?
```elixir
with_lock "my_key", fn ->
  do_something()
end
```

Reply:
That's a good point. The current version is a macro. I think the only good reason to keep it that way is to make it easier for people to update. That was @epinault's reasoning when I suggested passing a function instead of a do block.

Reply:
:shrug: I guess I feel like if this is dropping a 1.0 release it's a great time to hash out a better API

Reply:
I am ok with a breaking change. I would like to have a better API. The fact that already we are removing application.ex is big enough of a breaking change. So Agreed with @brettbeatty on this. Also I believe I suggested the same about passing an Anonymous function too. But having both would not hurt is what I also said. But ok to break the API too and support only anonymous function